kubevirt: add ImportProvider tab and WizardErrors component

suomiy opened the pull request:
```
 add ImportProvider tab
- move import redux state from VMSettings tab to its own
- refactor CreateVMWizard footer and implement new navigation logic + memoize redux
- add support for registering ws queries through redux (used by import provider updaters)
- add support for hiding tabs (through Edit button)
- add support for memory units and fix import (from template, v2v) edge cases
- add support for simple x advanced view
- allign names with DNS1123
- fix network and disk validation edge cases
- reorganize vmware provider and show statuses in the same place
- decouple functions from VMSettings tab
- set workoad to null if no relevant template found
- add simple tab validation for import provider
- make integer component more simpler and memoized
- deprecate IMPORT provision source - depends only on the
  isImportProvider prop

move createVM and disposeWizard actions into separate redux directory
- implement areStatePropsEqual for CreateVMWizard to have less rerenders
- add assertions for initial state
- unify usage of selectors
- create getStepsMetadata selector
- fix long values with units splitting
- kubevirt: improve and refactor ResourceLoadErrors

Implement goToStep functionality through redux
- don't change tab randomly when clicking on Edit
- move back button before review and create so the buttons don't
  accidentally switch when user is moving from the last tab

add WizardErrors component to the top of the wizard
- add actions to the errors - go to tab and open dialog
- v2vvmware import - set pod network by default when only 1 network is present
- deprecate getStoragesWithWrappers and getNetworkWithWrappers
    - to simplify code
    - remvove unnecesarry mappings when not needed
    - simplify interface of the wizard modals
- enhance CombinedDisk to receive also basic entities like disk/nic
```

depends on:
- [x] https://github.com/openshift/console/pull/4750/

suomiy commented:
please review @mareklibra @matthewcarleton @yaacov 

The design:

Import Tab

![review0](https://user-images.githubusercontent.com/3648838/77542887-29e9bf80-6ea7-11ea-8ffc-3d7b29b94c45.png)

Review Tab

![review1](https://user-images.githubusercontent.com/3648838/77542922-353ceb00-6ea7-11ea-92f0-b2e4c5e3a4de.png)

When you click on the network / disk error

![review2](https://user-images.githubusercontent.com/3648838/77542959-45ed6100-6ea7-11ea-9cf1-467c39d11fa2.png)

When you click on the network / disk field error

![review3](https://user-images.githubusercontent.com/3648838/77543014-543b7d00-6ea7-11ea-8edc-0f5b2d0ca2c3.png)

When you click on the VM attribute error

![review4](https://user-images.githubusercontent.com/3648838/77543081-6b7a6a80-6ea7-11ea-8c09-d16c59e9279e.png)

suomiy commented:
Tries to fix all names and sets the network to POD network if there is only one network.

Painless import (guessed and prefiled) 

![review6](https://user-images.githubusercontent.com/3648838/77543345-d9269680-6ea7-11ea-9bce-650743bb1235.png)

matthewcarleton commented:
@suomiy nice!
A few questions/comments:
Did you move the name/description to the first fields in the general? We updated this in the design now that we don't have import/template creation in the custom create flow we don't need to put these fields last.
When the import requires additional fields should we just drop them right into the general step (or whatever first step comes first) rather than require them to click through?
Can we remove the fields from the alert as they are completed?
I think the fields that are inside modals (nics, disks) should read like this " Networkadapter1 : Network" and be one link that sends them to the open modal. Same for disks.
When they hit the modal the field required should be highlighted with the error state.
For the footer actions we should have Next | Review and import | Back | Cancel
Can we switch the "Check and Save" action for the password to a secondary button? There should only ever be one primary action.
Thanks!

matthewcarleton commented on frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/error-components/wizard-errors.scss:
```diff
@@ -0,0 +1,16 @@
+.kv-create-vm-modal__wizard-errors {
+  width: 100%;
+  margin-bottom: 1em;
```
can we use a variable here

suomiy replied:
done

yaacov commented:
> Did you move the name/description to the first fields in the general? We updated this in the design now that we don't have import/template creation in the custom create flow we don't need to put these fields last.

@matthewcarleton hi, name and description is moving to the top in #4811

suomiy commented:
#4750 is merged -> rebased and we can continue with this PR

suomiy commented:
> When the import requires additional fields should we just drop them right into the general step (or whatever first step comes first) rather than require them to click through?

Currently, I made it so the user doesn't have to click edit and see all the steps immediately

I am not much of a fan of forcing actions onto the user automatically (changing steps). The user might want to see the review first or have a different order of fixing tabs in mind

But now the information about what is wrong is missing. I would rather add the errors header into the importers tab as well (didn't have it originally as you can switch VMs there)  

![rev1](https://user-images.githubusercontent.com/3648838/77673496-883b9e80-6f8a-11ea-9f8d-675a91e17183.png)



> Can we remove the fields from the alert as they are completed?

no bug there, already working like this

> I think the fields that are inside modals (nics, disks) should read like this " Networkadapter1 : Network" and be one link that sends them to the open modal. Same for disks.

Why do you prefer it this way?
I like the second way better, because it is more consistent and behaves like a navigation.

general_tab > general_field
nic_tab > nic
nic_tab > nic > nic_field

the fields have the same priority in the navigation for both the general tab and nic modal

> When they hit the modal the field required should be highlighted with the error state.

done

> For the footer actions we should have Next | Review and import | Back | Cancel

I found this combination quite frustrating when working with wizard

suppose you are in the review page  and wanna go 2 tabs back

`Next | Back | Cancel`

-> click  `Back` 

`Next | Review and import | Back | Cancel`

-> quick click in succession with a hope `Back` it will be there
-> mistakenly clicked `Review and import`
-> we are back in the review page

`Next | Back | Cancel`


> Can we switch the "Check and Save" action for the password to a secondary button? There should only ever be one primary action.

done

matthewcarleton commented:
> > When the import requires additional fields should we just drop them right into the general step (or whatever first step comes first) rather than require them to click through?
> 
> Currently, I made it so the user doesn't have to click edit and see all the steps immediately
> 
> I am not much of a fan of forcing actions onto the user automatically (changing steps). The user might want to see the review first or have a different order of fixing tabs in mind
> 
> But now the information about what is wrong is missing. I would rather add the errors header into the importers tab as well (didn't have it originally as you can switch VMs there)
> 
> ![rev1](https://user-images.githubusercontent.com/3648838/77673496-883b9e80-6f8a-11ea-9f8d-675a91e17183.png)
> 
> > Can we remove the fields from the alert as they are completed?
> 
> no bug there, already working like this
> 
> > I think the fields that are inside modals (nics, disks) should read like this " Networkadapter1 : Network" and be one link that sends them to the open modal. Same for disks.
> 
> Why do you prefer it this way?
> I like the second way better, because it is more consistent and behaves like a navigation.
> 
> general_tab > general_field
> nic_tab > nic
> nic_tab > nic > nic_field

 I think for the modal steps it's a bit odd to have to links (one to the NIC page, one to the NIC modal) They don't know where they error exists. If we only provide a link to the modal then they go to the right place.
What if we did it like this:
Nic_tab > Nic_name > Nic_field (the last one is the only link)
> the fields have the same priority in the navigation for both the general tab and nic modal
> 
> > When they hit the modal the field required should be highlighted with the error state.
> 
> done
> 
> > For the footer actions we should have Next | Review and import | Back | Cancel

When we designed it like this we had them in the order of actions that made the most sense, it's more likely the user wants to review than go back. It's an a11y thing too for tabbing order. I'm not against change here but I think we should do it separately b/c it's out of scope for this effort. WDYT?
> I found this combination quite frustrating when working with wizard
> 
> suppose you are in the review page and wanna go 2 tabs back
> 
> `Next | Back | Cancel`
> 
> -> click `Back`
> 
> `Next | Review and import | Back | Cancel`
> 
> -> quick click in succession with a hope `Back` it will be there
> -> mistakenly clicked `Review and import`
> -> we are back in the review page
> 
> `Next | Back | Cancel`
> 
> > Can we switch the "Check and Save" action for the password to a secondary button? There should only ever be one primary action.
> 
> done

suomiy commented:
> > > When the import requires additional fields should we just drop them right into the general step (or whatever first step comes first) rather than require them to click through?
> > 
> > 
> > Currently, I made it so the user doesn't have to click edit and see all the steps immediately
> > I am not much of a fan of forcing actions onto the user automatically (changing steps). The user might want to see the review first or have a different order of fixing tabs in mind
> > But now the information about what is wrong is missing. I would rather add the errors header into the importers tab as well (didn't have it originally as you can switch VMs there)

added the errors header to the importers tab as well - and it will be shown only after the vm has loaded

> What if we did it like this:
> Nic_tab > Nic_name > Nic_field (the last one is the only link)

ok sounds good, done
> When we designed it like this we had them in the order of actions that made the most sense, it's more likely the user wants to review than go back. It's an a11y thing too for tabbing order. I'm not against change here but I think we should do it separately b/c it's out of scope for this effort. WDYT?
> 

ok, reverted for now. It is more likely in the import flow - to for example edit the networks/disks

suomiy commented:
Added the SC warnings. Let me know if the messages could be improved.

![sc2](https://user-images.githubusercontent.com/3648838/77930757-93087300-72ab-11ea-8e75-eafc31728b3d.png)

![sc1](https://user-images.githubusercontent.com/3648838/77930767-96036380-72ab-11ea-9555-208b0531684d.png)

suomiy commented:
/test e2e-gcp-console

matthewcarleton commented:
> Added the SC warnings. Let me know if the messages could be improved.
> 
> SC not defined:
> 
> ![sc2](https://user-images.githubusercontent.com/3648838/77930757-93087300-72ab-11ea-8e75-eafc31728b3d.png)
> 
This looks good to me, do we want to add the default to the disk below?
> Default SC missing:
> 
> ![sc1](https://user-images.githubusercontent.com/3648838/77930767-96036380-72ab-11ea-9555-208b0531684d.png)

Won't it always be the case where all the disks won't have a SC? If that's the case then using "some" feels inaccurate. 
What happens when they click "create a default one"?

suomiy commented:
> > SC not defined:
> 
> This looks good to me, do we want to add the default to the disk below?

Not sure, we are keeping it empty by k8s convention as it will get mapped to the default one on creation.

> 
> > Default SC missing:
>
> Won't it always be the case where all the disks won't have a SC? If that's the case then using "some" feels inaccurate.

No, some might have it, and some might not.

> What happens when they click "create a default one"?

They will be redirected to the storageclasses page.  It is up to the user to decide if s/he wants to create a new one, or mark existing one as a default one.

matthewcarleton commented:
> > > SC not defined:
> > 
> > 
> > This looks good to me, do we want to add the default to the disk below?
> 
> Not sure, we are keeping it empty by k8s convention as it will get mapped to the default one on creation.
> 
> > > Default SC missing:
> > 
> > 
> > Won't it always be the case where all the disks won't have a SC? If that's the case then using "some" feels inaccurate.
> 
> No, some might have it, and some might not.
> 
ok, can you explain this a bit more. I thought if VMware was the provider there was no chance a storage class would be imported and we could only rely on already set defaults? I'm sure I'm missing something just want to make sure this covers every case.

> > What happens when they click "create a default one"?
> 
> They will be redirected to the storageclasses page. It is up to the user to decide if s/he wants to create a new one, or mark existing one as a default one.

Will this open in a new tab? That would be ideal so they don't lose what they've done. Although we'd have to be able to refresh to see the newly assigned default class.

suomiy commented:
> ok, can you explain this a bit more. I thought if VMware was the provider there was no chance a storage class would be imported and we could only rely on already set defaults? I'm sure I'm missing something just want to make sure this covers every case.

It might be, that some storage classes exist, but none of them is marked as default. For some storages the user might have picked the storage classes, and for some the user might not.

> > 
> > They will be redirected to the storageclasses page. It is up to the user to decide if s/he wants to create a new one, or mark existing one as a default one.
> 
> Will this open in a new tab? That would be ideal so they don't lose what they've done. Although we'd have to be able to refresh to see the newly assigned default class.

You are asked if you want to throw away the work done in the wizard and on yes it will redirect in the same window. The user has also an option to middle/right click on the link and open it in the new window. IMO it is good enough for now, but we can improve on that later. 

If the user marks/creates the default SC in a second window, then the changes should be reflected immediately

suomiy commented:
rebased

suomiy commented:
/test e2e-gcp-console

suomiy commented:
/test e2e-gcp-console

suomiy commented:
/test e2e-gcp-console

suomiy commented:
rebased

suomiy commented:
/test e2e-gcp-console

yaacov commented:
/test e2e-gcp-console

yaacov commented:
/lgtm
